pytorch_lightning/utilities/cli.py

```diff
@@ -153,107 +156,154 @@ def __init__(
         self.datamodule_class = datamodule_class
         self.save_config_callback = save_config_callback
         self.trainer_class = trainer_class
-        self.trainer_defaults = {} if trainer_defaults is None else trainer_defaults
+        self.trainer_defaults = trainer_defaults or {}
+        #self.trainer_fn = trainer_fn
         self.seed_everything_default = seed_everything_default
         self.subclass_mode_model = subclass_mode_model
         self.subclass_mode_data = subclass_mode_data
-        self.parser_kwargs = {} if parser_kwargs is None else parser_kwargs
-        self.parser_kwargs.update({'description': description, 'env_prefix': env_prefix, 'default_env': env_parse})
 
-        self.init_parser()
-        self.add_core_arguments_to_parser()
-        self.add_arguments_to_parser(self.parser)
-        self.parse_arguments()
+        parser_kwargs = parser_kwargs or {}
+        parser_kwargs.update({'description': description, 'env_prefix': env_prefix, 'default_env': env_parse})
+        self.parser = self.init_parser(**parser_kwargs)
+
+        self.add_core_arguments(self.parser)
+        self.add_arguments(self.parser)
+        self.add_subcommands(self.parser)
```

Comment:
`add_arguments` since is the method that people extend should be last. For example `link_arguments` requires the arguments to already exist to be able to link them.

```suggestion
        self.add_subcommands(self.parser)
        self.add_arguments(self.parser)
```